internal/server/data/group.go

```diff
@@ -41,19 +40,37 @@ func ListGroups(db GormTxn, p *Pagination, selectors ...SelectorFunc) ([]models.
 	}
 
 	return groups, nil
-
 }
 
 func ByGroupMember(id uid.ID) SelectorFunc {
 	return func(db *gorm.DB) *gorm.DB {
 		return db.
-			Joins("JOIN identities_groups ON groups.id = identities_groups.group_id").
-			Where("identities_groups.identity_id = ?", id)
+			Joins(`
+			JOIN (
+				-- Combine the direct group members and the members of mapped identity provider groups
+				SELECT * FROM identities_groups
+				NATURAL FULL JOIN (
+					-- See identities in groups based on their provider group mappings
+					SELECT identity_id, group_id from provider_groups_identities
+					JOIN provider_groups_groups ON provider_groups_groups.provider_group_id = provider_groups_identities.provider_group_id
+				) AS normalized_identity_groups
+				) AS resolved_identity_groups
+			ON resolved_identity_groups.group_id = id
+			`).
+			Where("resolved_identity_groups.identity_id = ?", id)
 	}
 }
 
 func groupIDsForUser(tx ReadTxn, userID uid.ID) ([]uid.ID, error) {
-	stmt := `SELECT DISTINCT group_id FROM identities_groups WHERE identity_id = ?`
+	stmt := `
+		SELECT DISTINCT group_id FROM identities_groups
+		NATURAL FULL JOIN (
+			-- See identities in groups based on their provider group mappings
+			SELECT identity_id, group_id from provider_groups_identities
+			JOIN provider_groups_groups ON provider_groups_groups.provider_group_id = provider_groups_identities.provider_group_id
+		) AS normalized_identity_groups
+		WHERE identity_id = ?
```

Comment:
I'm hoping we can remove some of these full joins and sub-selects.

Would something like this work?

```sql
SELECT DISTINCT provider_groups.group_id
FROM provider_groups_identities
LEFT JOIN provider_groups ON provider_groups_identities.provider_group_id = provider_groups.id
WHERE provider_groups_identities.identity_id = ?
```

Reply:
I definitely want to simplify these, but even more I want to remove them completely by making the sync do the work of updating the memberships.